lib/test.js

```diff
@@ -550,14 +551,28 @@ Test.prototype._runChild = function runChild (child, name, extra, cb) {
 
   // still need try/catch for synchronous errors
   try {
+    child._mustDeferEnd = true;
     var cbRet = cb(child)
     if (cbRet && typeof cbRet.then === 'function') {
       // promise
       cbRet.then(function () {
-        child.end(IMPLICIT)
-      }, child.threw)
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
+          child.end(IMPLICIT)
+        }
+      }, function (err) {
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
```

Comment:
It doesn't seems to affect the test results if this is changed to unconditionally call `child.threw(err)`. Missing test cases?